src/lib.rs

```diff
@@ -1,14 +1,84 @@
-mod totp;
-mod hotp;
+use hmac::{
+    crypto_mac::InvalidKeyLength,
+    Hmac, Mac, NewMac,
+};
+use sha1::Sha1;
 
-pub fn digest() {}
-pub fn generate_secret(){}
-pub fn get_otp_auth_url(){}
+type HmacSha1 = Hmac<Sha1>;
+
+pub mod hotp;
+pub mod totp;
+
+enum HmacFunction<A> {
+    Sha1(A),
+}
+
+pub enum Algorithm {
+    Sha1,
+    // Todo (Jayhelton)Implement other hash functions
+    // Sha256,
+    // Sha512,
+}
+
+// pub enum Encoding {
+//     Ascii,
+//     Hex,
+//     Base32,
+//     Base64,
+// }
+
+pub fn digest(
+    secret: String,
+    counter: u128,
+    algorithm: Algorithm,
+) -> std::result::Result<Vec<u8>, InvalidKeyLength> {
+    let mac = match get_hmac(secret, algorithm) {
+        Ok(_hmac) => _hmac,
+        Err(e) => return Err(e),
+    };
+
+    match mac {
+        HmacFunction::Sha1(mut _mac) => {
+            _mac.update(b"input mdddessage");
+            Ok(_mac.finalize().into_bytes().to_vec())
+        }
+    }
+}
+
+pub fn generate_secret() {}
+pub fn get_otp_auth_url() {}
+
+fn get_hmac(
+    secret: String,
+    algorithm: Algorithm,
+) -> Result<HmacFunction<HmacSha1>, InvalidKeyLength> {
```

Comment:
I imagine that you are going to make this return a result of the different HmacFunction<HASH_FUNCTION> when you get to the other hash algos?